### PR TITLE
Res 1594

### PR DESCRIPTION
**Pull request by karangrewal**
Multi-head setup taken from the continual learning benchmarks repository. This code is compatible with all types of heads, and splitMNIST and splitGSC.

https://github.com/GMvandeVen/continual-learning

### CODE REVIEW

File: nupic/research/frameworks/continuous_learning/multihead/multihead.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+                print("\ttask {}/{}: {}".format(i + 1, n_tasks, results))
+
+            print("Epoch {}: {}".format(epoch, results))
+
+
+def get_active_classes(task_num, scenario):
+    """
+    Return a list of label indices that are "active" during training under the given
+    scenario
+    :param task_num: the index of the task
```
Comment (lucasosouza) [inline nupic/research/frameworks/continuous_learning/multihead/multihead.py:189]: Can you elaborate a little bit on the reasoning behind the <task, class, domain> heuristic defined below? This should be somewhere in the documentation. Here is a good place to do it, the `subdataset.get_datasets` method is another alternative.

File: nupic/research/frameworks/continuous_learning/multihead/subdataset.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+from torchvision.datasets import MNIST
+
+
+def get_datasets(dataset_name, scenario, download=False):
+    """
+    Return the train and test datasets (in parallel lists) for each task in the
+    continuous learning setup
+    :param dataset_name: name of the dataset to retrieve
+    :type dataset_name: str
+    :param scenario: continuous learning setup, one of {'task', 'domain', 'class'}
```
Comment (lucasosouza) [inline nupic/research/frameworks/continuous_learning/multihead/subdataset.py:41]: This function doesn't distinguish between task or class, it only cares whether it is domain or not, which could be represented as a boolean flag to simplify the interface.

File: nupic/research/frameworks/continuous_learning/multihead/subdataset.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    dataset_transform = transforms.Compose([
+        transforms.ToTensor(),
+    ])
+
+    if dataset_name == "splitGSC":
+        dataset_class = GSC
+    elif dataset_name == "splitMNIST":
+        dataset_class = MNIST
+
+    full_train_set = dataset_class("~/nta/continual-learning/datasets/mnist/",
```
Comment (lucasosouza) [inline nupic/research/frameworks/continuous_learning/multihead/subdataset.py:57]: I suggest to make the path to the dataset relative (possibly a command line argument in `train_model` or an environment variable), it will make it easier for others to run.

File: nupic/research/frameworks/continuous_learning/multihead/subdataset.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+                                        target_transform=target_transform))
+
+    return train_datasets, test_datasets
+
+
+class GSC(Dataset):
+    """ Google Speech Commands dataset. """
+
+    def __init__(self, root=".", train=True, transform=None, target_transform=None,
+                 download=False):
```
Comment (lucasosouza) [inline nupic/research/frameworks/continuous_learning/multihead/subdataset.py:85]: This download flag is not being used. If it is here to maintain consistency with the torchvision API, it should throw an exception if the dataset is not available, warning the user that the download flag is being ignored and the dataset has to be manually downloaded.

File: nupic/research/frameworks/continuous_learning/multihead/subdataset.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+            # remove items with labels 0 or 1
+            idx = (self.test_labels > 1)
+            self.data = self.data[idx]
+            self.test_labels = self.test_labels[idx]
+
+            # shift all labels to be at 0
+            self.test_labels = self.test_labels - 2
+
+            # select subset of data
```
Comment (lucasosouza) [inline nupic/research/frameworks/continuous_learning/multihead/subdataset.py:134]: Best to remove commented code, if not relevant anymore

File: nupic/research/frameworks/continuous_learning/multihead/multihead.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+    :param device:
+    :type device: :class:`torch.device`
+    :param post_batch_callback: function(model) to call after every batch
+    :type post_batch_callback: function
+    """
+    model.train()
+    for data, target in tqdm(loader, desc="Train", leave=False):
+        data, target = data.to(device), target.to(device)
+        if active_classes is not None:
+            target = target - active_classes[0]
```
Comment (lucasosouza) [inline nupic/research/frameworks/continuous_learning/multihead/multihead.py:68]: There is a similar transformation applied as a target_transform in `subdataset.get_datasets`. What is the difference, why is in that case applied directly to the dataset, while here is applied in the training function after the target is sampled from the dataloader?

File: projects/continuous_learning/multihead/model.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        if is_sparse:
+            self.fc1 = SparseWeights(self.fc1, sparsity=0.75)
+            self.kw1 = KWinners(n=n_hidden_units, percent_on=0.1, boost_strength=0.0)
+
+            self.fc2 = SparseWeights(self.fc2, sparsity=0.85)
+            self.kw2 = KWinners(n=n_hidden_units, percent_on=0.1, boost_strength=0.0)
+
+    def forward(self, x):
+        output = self.fc1(self.flatten(x))
+        output = self.kw1(output) if self.is_sparse else output
```
[the remark this answers is not in the document]
Comment (karangrewal) [reply projects/continuous_learning/multihead/model.py:54]: good catch, I'll add ReLU.

File: nupic/research/frameworks/continuous_learning/multihead/subdataset.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+
+class SubDataset(Dataset):
+    """ A subdataset class that divides a given dataset with labels into one with
+    specified labels. """
+
+    def __init__(self, original_dataset, sub_labels, target_transform=None):
+        super().__init__()
+        self.dataset = original_dataset
+        self.sub_indeces = []
```
[the remark this answers is not in the document]
Comment (karangrewal) [reply nupic/research/frameworks/continuous_learning/multihead/subdataset.py:163]: fixed.

File: nupic/research/frameworks/continuous_learning/multihead/multihead.py (hunk 1 above)
Comment (karangrewal) [reply nupic/research/frameworks/continuous_learning/multihead/multihead.py:189]: I updated the docstring to include further explanations and a link to the continual learning benchmarks paper.

File: nupic/research/frameworks/continuous_learning/multihead/subdataset.py (hunk 2 above)
Comment (karangrewal) [reply nupic/research/frameworks/continuous_learning/multihead/subdataset.py:41]: I made the change you suggested, and modified all other modules that call this function.

File: nupic/research/frameworks/continuous_learning/multihead/subdataset.py (hunk 3 above)
Comment (karangrewal) [reply nupic/research/frameworks/continuous_learning/multihead/subdataset.py:57]: I'll just use the root directory.

File: nupic/research/frameworks/continuous_learning/multihead/subdataset.py (hunk 4 above)
Comment (karangrewal) [reply nupic/research/frameworks/continuous_learning/multihead/subdataset.py:85]: Noted.

File: nupic/research/frameworks/continuous_learning/multihead/subdataset.py (hunk 5 above)
Comment (karangrewal) [reply nupic/research/frameworks/continuous_learning/multihead/subdataset.py:134]: Done.

File: nupic/research/frameworks/continuous_learning/multihead/subdataset.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+                    label = self.dataset.test_labels[index]
+                else:
+                    label = self.dataset.target_transform(
+                        self.dataset.test_labels[index]
+                    )
+            else:
+                label = self.dataset[index][1]
+            if label in sub_labels:
+                self.sub_indeces.append(index)
+        self.target_transform = target_transform
```
[the remark this answers is not in the document]
Comment (karangrewal) [reply nupic/research/frameworks/continuous_learning/multihead/subdataset.py:181]: This code is mostly taken from the benchmarks repo (link here), but looks like they are only applying the dataset's target_transform within the loop, which is separate than the target_transform that is a parameter to the __init__ function here. This target_transform is applied to create subdatasets, e.g., if you want to always have labels [0, 1] in the case where there is a single output head, the subdataset's target_transform would take care of that, not the dataset's target_transform.

[the remark this answers is not in the document]
Comment (karangrewal) [reply nupic/research/frameworks/continuous_learning/multihead/subdataset.py:181]: Furthermore, the subdataset's target_transform is only defined after the loop since it only gets used in the __getitem__ function.

File: nupic/research/frameworks/continuous_learning/multihead/multihead.py (hunk 6 above)
Comment (karangrewal) [reply nupic/research/frameworks/continuous_learning/multihead/multihead.py:68]: You're right, I placed all of it in the subdataset class.

File: projects/continuous_learning/multihead/model.py [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+Dense/sparse MLP implementation
+"""
+
+import torch
+import torch.nn.functional as F
+
+from nupic.torch.modules import Flatten, KWinners, SparseWeights
+
+
+class Classifier(torch.nn.Module):
```
Comment (subutai) [inline projects/continuous_learning/multihead/model.py:32]: I think we should have a more specific name than `Classsifier`, which could mean almost anything. Maybe `TwoLayerClassifier` or something like that?

File: projects/continuous_learning/multihead/model.py [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+        self.is_sparse = is_sparse
+        self.flatten = Flatten()
+
+        self.fc1 = torch.nn.Linear(input_size, n_hidden_units)
+        self.fc2 = torch.nn.Linear(n_hidden_units, n_hidden_units)
+        self.fc3 = torch.nn.Linear(n_hidden_units, n_classes)
+
+        if is_sparse:
+            self.fc1 = SparseWeights(self.fc1, sparsity=0.75)
```
Comment (subutai) [inline projects/continuous_learning/multihead/model.py:47]: All these parameters such as `sparsity`, `percent_on`, etc. should be parameters to the `__init__` instead of hardcoded constants.

File: nupic/research/frameworks/continuous_learning/multihead/multihead.py [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+import torch
+import torch.nn.functional as F
+import torch.optim as optim
+from tqdm import tqdm
+
+from nupic.research.frameworks.continuous_learning.multihead.subdataset import (
+    get_datasets,
+)
+
+SEED = 42
```
[the remark this answers is not in the document]
Comment (karangrewal) [reply nupic/research/frameworks/continuous_learning/multihead/multihead.py:39]: you are correct; it's not used.  I'll remove this variable.

File: nupic/research/frameworks/continuous_learning/multihead/multihead.py [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+                                    batch_size=TEST_BATCH_SIZE,
+                                    shuffle=True)
+        for test_dataset in test_datasets]
+
+    n_tasks = len(train_datasets)
+
+    # optimizer for training model
+    adam = optim.Adam(model.parameters(), lr=LEARNING_RATE, betas=(0.9, 0.999))
+
+    for epoch in range(EPOCHS):
```
[the remark this answers is not in the document]
Comment (karangrewal) [reply nupic/research/frameworks/continuous_learning/multihead/multihead.py:162]: Fixed.

File: projects/continuous_learning/multihead/model.py (hunk 10 above)
Comment (karangrewal) [reply projects/continuous_learning/multihead/model.py:32]: Sounds good. I also added documentation to the init function.

File: projects/continuous_learning/multihead/model.py (hunk 11 above)
Comment (karangrewal) [reply projects/continuous_learning/multihead/model.py:47]: Done.

**Comment by karangrewal** [comment]
I made the changes that you requested @subutai and updated this PR. The tests are not included (will have them shortly) so I can merge the 4 modules now or afterwards, once I add the tests.
